### PR TITLE
Multishop search mechanism for header

### LINKED ISSUE #19300: Multistore - Add new header on already migrated pages
As a merchant I want to be able to switch between contexts when I have many stores, so I can choose the store(s) I want to configure and easily see which store I am configuring 

[Specs header](https://github.com/PrestaShop/prestashop-specs/blob/master/back-office/multistoregeneralspecs.md#multistore-header-only-from-178)

This header should be displayed on all pages already migrated

As specified in the [multistore specials cases specs](https://github.com/PrestaShop/prestashop-specs/blob/master/broader-topics/multistorespecialsspecs.md#list-of-pages-for-specific-case-3-the-page-can-be-displayed-only-in-all-shops-context), some pages can be displayed only in "all shops" context and the others contexts are not displayed in the header

List of theses pages:
- Shipping > Carriers > Add new carrier
- International > Localization > Languages
- Shop parameters > Order settings > Statuses
- Shop parameters > Order settings > Statuses > Add new order status & add new order return status
- Advanced parameters > Team > Profiles
- Advanced parameters > Team > Permissions
- Advanced parameters > Multistore > AdminShopGroup
- Advanced parameters > Multistore > AdminShop
- Advanced parameters > Multistore > AdminShopUrl
- Advanced parameters > Multistore > Add new shop group
- Advanced parameters > Multistore > Add new shop
- Advanced parameters > Multistore > Add new URL

### PR DESCRIPTION
**Pull request by matthieu-rolland**
| Questions     | Answers
| ------------- | -------------------------------------------------------
| Branch?       | develop
| Description?  | In the multishop header, there will be a field to search for shops. This PR implements the related query mechanism, it also initiate the new shop controller and shop behat tests.
| Type?         | new feature
| Category?     | CO
| BC breaks?    | no
| Deprecations? | no
| Fixed ticket? | Partially fixes #19300
| How to test?  | **In BO:**<br><br>1/ Activate multishop<br> 2/ Have two or more shops<br> 3/ With your browser or postman, call this url: `/configure/advanced/shops/search/yourshop` where "yourshop" is a a part of the name of an existing shop.<br><br> If the shop exists you should receive a json with shop information (name etc...), if the shop doesn't exist, you will get an empty json and a 404, if you omit the search parameter you should get a 404 as well.

---
This change is [Reviewable](https://reviewable.io/reviews/prestashop/prestashop/21593)

### CODE REVIEW

File: src/PrestaShopBundle/Entity/Repository/ShopRepository.php
Comment (Progi1984) [inline src/PrestaShopBundle/Entity/Repository/ShopRepository.php]:
```suggestion
    public function findBySearchTerm(string $searchTerm): array
```

**Review by Progi1984 — COMMENTED** [review]
A small feedback

**Review by Progi1984 — DISMISSED** [review]

**Comment by matthieu-rolland** [comment]
@Progi1984 I have a fail with behat, I'm dealing with it !

File: src/Core/Domain/Shop/Query/SearchShops.php
Comment (jolelievre) [inline src/Core/Domain/Shop/Query/SearchShops.php]: I think you can create a more precise exception `SearchShopException` that extends `ShopException`

File: src/PrestaShopBundle/Controller/Admin/Configure/AdvancedParameters/ShopController.php
Comment (jolelievre) [inline src/PrestaShopBundle/Controller/Admin/Configure/AdvancedParameters/ShopController.php]: If `$shops` is empty maybe the HTTP code should be 404 (but still return a JSON content, even if empty)

File: src/PrestaShopBundle/Entity/Repository/ShopRepository.php
Comment (jolelievre) [inline src/PrestaShopBundle/Entity/Repository/ShopRepository.php]: Maybe you should use `LOWER(s.name)` and `'%' . strtolower($searchTerm) . '%'` So that your search is not case sensitive

From what I saw on a stackoverflow message you might need to do `$qb->expr()->like('LOWER(s.name)', '?1'),` for the LOWER to work correctly 🤷‍♂️  (but it was in 2013 😅 )

File: tests/Integration/Behaviour/Features/Scenario/Shop/search_shops.feature
Comment (jolelievre) [inline tests/Integration/Behaviour/Features/Scenario/Shop/search_shops.feature]: I don't think you need to split into two steps, it forces you to store the result which is prone to errors
You can simply do:
```
When I search for shops with the term "test" I get following results:
      | name             | group_name |
      | test_shop        | Default    |
      | test_second_shop | Default    |
      | test_third_shop  | Default    |
```

File: src/PrestaShopBundle/Entity/Repository/ShopRepository.php
Comment (matthieu-rolland) [reply src/PrestaShopBundle/Entity/Repository/ShopRepository.php]: done !

File: src/Core/Domain/Shop/Exception/SearchShopException.php
Comment (Progi1984) [inline src/Core/Domain/Shop/Exception/SearchShopException.php]:
```suggestion
 */

declare(strict_types=1);

namespace PrestaShop\PrestaShop\Core\Domain\Shop\Exception;
```

File: tests/Integration/Behaviour/Features/Context/ShopFeatureContext.php
Comment (Progi1984) [inline tests/Integration/Behaviour/Features/Context/ShopFeatureContext.php]:
```suggestion
    public const LAST_FOUND_SHOPS_KEY = 'LAST_FOUND_SHOPS';
```

**Review by Progi1984 — COMMENTED** [review]
Small feedbacks

**Comment by Progi1984** [comment]
@matthieu-rolland Don't forget to squash your commits

**Review by Progi1984 — DISMISSED** [review]

File: app/config/config_test.yml
Comment (NeOMakinG) [inline app/config/config_test.yml]: Need new line at the end of this file

**Review by Progi1984 — DISMISSED** [review]

File: app/config/config_test.yml
Comment (matthieu-rolland) [reply app/config/config_test.yml]: done !

**Review by Progi1984 — DISMISSED** [review]

File: src/Core/Domain/Shop/QueryHandler/SearchShopsHandlerInterface.php
Comment (sowbiba) [inline src/Core/Domain/Shop/QueryHandler/SearchShopsHandlerInterface.php]: GetLogos ?

File: src/PrestaShopBundle/Controller/Admin/Configure/AdvancedParameters/ShopController.php
Comment (matks) [inline src/PrestaShopBundle/Controller/Admin/Configure/AdvancedParameters/ShopController.php]: If you use this in every BO page you will need to make it available to all

**Review by sowbiba — COMMENTED** [review]
Good job. A small comment. Sorry dismissing the approval

File: src/PrestaShopBundle/Resources/config/routing/admin/configure/advanced_parameters/shops.yml
Comment (matks) [inline src/PrestaShopBundle/Resources/config/routing/admin/configure/advanced_parameters/shops.yml:5]: I think you need a `_legacy_controller` for some of the stuff to work 🤔 
- tabs
- authorization

ping @jolelievre @zuk3975

Comment (matthieu-rolland) [reply src/PrestaShopBundle/Resources/config/routing/admin/configure/advanced_parameters/shops.yml:5]: I was unsure if it was needed.... since there was no corresponding "search" method or similar in the legacy AdminShopController...  (it's a new feature)

**Review by sowbiba — DISMISSED** [review]

File: src/PrestaShopBundle/Controller/Admin/Configure/AdvancedParameters/ShopController.php
Comment (matthieu-rolland) [reply src/PrestaShopBundle/Controller/Admin/Configure/AdvancedParameters/ShopController.php]:
```suggestion
     * @AdminSecurity("is_granted('read', request.get('_legacy_controller'))")
```

Did you mean something like this? (and setting `_legacy_controller` in the route definition)

File: src/Adapter/Shop/QueryHandler/SearchShopsHandler.php
Comment (jolelievre) [inline src/Adapter/Shop/QueryHandler/SearchShopsHandler.php:65]: Usually CQRS query return container objects instead of array
And then if the data needs to be formatted as array it is handled in the controller for example

File: tests/Integration/Behaviour/Features/Context/ShopFeatureContext.php
Comment (jolelievre) [inline tests/Integration/Behaviour/Features/Context/ShopFeatureContext.php]: This function is now useless It's a good thing, storing the result of a step to test it later is a bit dangerous as the whole test becomes stateful

**Review by jolelievre — DISMISSED** [review]
Query should use container objects not plain arrays
And you can clear a useless function in tests (along with the now useless const)

File: tests/Integration/Behaviour/Features/Context/ShopFeatureContext.php
Comment (jolelievre) [inline tests/Integration/Behaviour/Features/Context/ShopFeatureContext.php]: Why do you need to compare with lower keys? You could check strict equality to make sure the real shop name is returned no?

File: tests/Integration/Behaviour/Features/Scenario/Shop/search_shops.feature
Comment (jolelievre) [inline tests/Integration/Behaviour/Features/Scenario/Shop/search_shops.feature]: You make no tests about color?

File: tests/Integration/Behaviour/Features/Context/ShopFeatureContext.php
Comment (matthieu-rolland) [reply tests/Integration/Behaviour/Features/Context/ShopFeatureContext.php]: Nice catch, it's a mistake. (I compare lower keys when I compare user input, but that's not the case here)

File: tests/Integration/Behaviour/Features/Scenario/Shop/search_shops.feature
Comment (jolelievre) [inline tests/Integration/Behaviour/Features/Scenario/Shop/search_shops.feature:18]: I feels weird to have empty colors in this test But the lines after you actually check that there are colors
In the internals of checking you just ignore when color is empty but it can be a valid test (if no color is set for example) So you should correctly set all the colors in all tests so that they remain consistent (especially if the value is empty)

File: src/Core/Domain/Shop/QueryHandler/SearchShopsHandlerInterface.php
Comment (Progi1984) [inline src/Core/Domain/Shop/QueryHandler/SearchShopsHandlerInterface.php]:
```suggestion

```

File: tests/Integration/Behaviour/Features/Scenario/Shop/search_shops.feature
Comment (Progi1984) [inline tests/Integration/Behaviour/Features/Scenario/Shop/search_shops.feature]:
```suggestion
      | test_shop        | Default    |       |
      | test_second_shop | Default    | red   |
      | test_third_shop  | Default    | blue  |
```

**Review by Progi1984 — COMMENTED** [review]
Small feedback

File: src/Core/Domain/Shop/QueryHandler/SearchShopsHandlerInterface.php
Comment (zuk3975) [inline src/Core/Domain/Shop/QueryHandler/SearchShopsHandlerInterface.php]: missing return type

File: src/Core/Domain/Shop/QueryHandler/SearchShopsHandlerInterface.php
Comment (zuk3975) [inline src/Core/Domain/Shop/QueryHandler/SearchShopsHandlerInterface.php]: you can add `@see SearchShops` to typehint the query, so it would be recognized by IDE (like when renaming the query class docblock wouldn't get outdated).
And ( optional-taste preference) I like to comment it like `Defines contract to handle @see SearchShops query` instead of repeating the "interface" and "service"

File: tests/Integration/Behaviour/Features/Context/ShopFeatureContext.php
Comment (zuk3975) [inline tests/Integration/Behaviour/Features/Context/ShopFeatureContext.php]:
```suggestion
    public function addShopGroup(string $reference, string $groupName, string $color): void
```

File: tests/Integration/Behaviour/Features/Context/ShopFeatureContext.php
Comment (zuk3975) [inline tests/Integration/Behaviour/Features/Context/ShopFeatureContext.php]: FYI if each table row represents FoundShop DTO then you can cast it using `@Transform`
like this https://github.com/PrestaShop/PrestaShop/pull/22244/files#diff-82fec04e223dab8f56f4604efd062e83a9058da2851559d55c88ca0f902d20c6R63-R102

Comment (matthieu-rolland) [reply tests/Integration/Behaviour/Features/Context/ShopFeatureContext.php]: hmm, interesting, I'll have a look 👍

Comment (matthieu-rolland) [reply tests/Integration/Behaviour/Features/Context/ShopFeatureContext.php]: Ok, I started doing it, but the problem in my case is that FoundShop is an existing shop, it takes a shop ID and a group shop ID in the constructor, and since IDs might change depending on the tests executed before in the same feature suite, I don't want to rely on them for anything... nor do I like the idea of putting a fake/random ID just for the test.

Comment (matthieu-rolland) [reply tests/Integration/Behaviour/Features/Context/ShopFeatureContext.php]: But I didn't know about this @transform keyword, I'll definitely use it when possible.

Comment (zuk3975) [reply tests/Integration/Behaviour/Features/Context/ShopFeatureContext.php]: > it takes a shop ID and a group shop ID in the constructor, and since IDs might change depending on the tests executed before in the same feature

In theory each step should be independent from each other (we are far from it though :sweat: ).
Using references to reflect ids would work if Background contains only "exists" assertion (when shop is created only if it does not exist yet) whilst now for each scenario it is creating a new shop while keeping same reference (that seems to be wrong :sweat:). 
E.g. if you add another scenario, then your shops would be duplicated by the background and you would expect to find multiple shops with same names and same references - it doesn't seem right.

I would suggest simply to move following lines from background to scenario
```
    Given I add a shop group "shopGroup2" with name "test_second_shop_group" and color "green"
    And I add a shop "shop2" with name "test_second_shop" and color "red" for the group "test_second_shop_group"
    And I add a shop "shop3" with name "test_third_shop" and color "blue" for the group "test_second_shop_group"
```
Then there won't be any problem adding another independant scenario and also you could take advantage of transformers by using references for shopId and groupId.
Im not very confident with this as actually somehow we should encourage using background :sweat:  so I ping @jolelievre, @matks to check this out (maybe have some discussion in slack), at least to see the problem with backgrounds

File: src/Adapter/Shop/QueryHandler/SearchShopsHandler.php
Comment (atomiix) [inline src/Adapter/Shop/QueryHandler/SearchShopsHandler.php]:
```suggestion
```

**Review by Progi1984 — DISMISSED** [review]

File: tests/Integration/Behaviour/Features/Scenario/Shop/search_shops.feature
Comment (zuk3975) [inline tests/Integration/Behaviour/Features/Scenario/Shop/search_shops.feature]: I suspect, this means empty results.
Its worth to add dedicated method for this with something like:
`When I search for shops with the term "doesnt_exist" I should not get any results`
It is much more readable.

File: src/Adapter/Shop/QueryHandler/SearchShopsHandler.php
Comment (PierreRambaud) [inline src/Adapter/Shop/QueryHandler/SearchShopsHandler.php]:
```suggestion
                $shop['color'] ?? '',
```

File: src/Adapter/Shop/QueryHandler/SearchShopsHandler.php
Comment (Progi1984) [reply src/Adapter/Shop/QueryHandler/SearchShopsHandler.php]: PR #20125 has been merged. Please remove the comment

**Review by Progi1984 — COMMENTED** [review]
A small feedback

File: tests/Integration/Behaviour/Features/Scenario/Shop/search_shops.feature
Comment (matthieu-rolland) [reply tests/Integration/Behaviour/Features/Scenario/Shop/search_shops.feature:18]: now I check colors all the time (even when it's empty), but maybe (probably) it was not the case when you wrote this feedback

**Review by Progi1984 — DISMISSED** [review]

**Review by matks — DISMISSED** [review]

**Review by jolelievre — DISMISSED** [review]

**Comment by SD1982** [comment]
Hi @matthieu-rolland 
as we saw together I did not have the expected result
https://drive.google.com/file/d/1JHs432ADHlBzpu_9bqQu0Q4K5Q7Tkut9/view?usp=sharing

**Review by matks — DISMISSED** [review]

**Review by Progi1984 — DISMISSED** [review]

**Comment by SD1982** [comment]
Hi @matthieu-rolland i just have a page refresh when i modify the url
https://drive.google.com/file/d/1k2rZyL_wAP8TWOraINhWxbVDYrTIXRCd/view?usp=sharing
Did I miss something ?

File: tests/Integration/Behaviour/Features/Context/ShopFeatureContext.php
Comment (jolelievre) [inline tests/Integration/Behaviour/Features/Context/ShopFeatureContext.php:96]: You will probably have conflicts with this pending PR from 177 https://github.com/PrestaShop/PrestaShop/pull/22228

Either you or when we merge it to develop, maybe we should wait for it to be merged in 177 and then back to develop, it will be easier to solve conflicts on this PR than during the merge which will lack context

Comment (jolelievre) [reply tests/Integration/Behaviour/Features/Context/ShopFeatureContext.php:96]: Nevermind it's the same code copied anyway ^^

**Review by jolelievre — APPROVED** [review]

**Review by matks — APPROVED** [review]

**Comment by matthieu-rolland** [comment]
> Hi @matthieu-rolland i just have a page refresh when i modify the url
> https://drive.google.com/file/d/1k2rZyL_wAP8TWOraINhWxbVDYrTIXRCd/view?usp=sharing
> Did I miss something ?

should be fine now, same URL as before 👍

**Review by PierreRambaud — APPROVED** [review]

**Comment by SD1982** [comment]
LGTM !!
Tested with 5 differents shops
Shops with similar root names
Shops with differents names
Shop with mixed name
Shop with number in the name
Shop with number in the name and whose numbers could be confused with the shop id when searching (ex demo12345)
Search with 1, 2, 3, 4 characters
Search with 1, 2, 3, 4 numbers
Search with non-existent shop name
Search with non-existent number
Search with empty search parameter

**Some randoms screenshots**
image
image
image
image
image
image
image
image
image
image

**Comment by matthieu-rolland** [comment]
thank you @SD1982 !

**Comment by Progi1984** [comment]
Great news. Thans @SD1982 & @matthieu-rolland

**Pull request merged**
